src/ui/win32/Dlg_MessageBox.cpp

```diff
@@ -0,0 +1,121 @@
+
+#include "RA_Defs.h" // for NativeStr
+
+#include "ui/viewmodels/MessageBoxViewModel.hh"
+
+#include <WTypes.h>
+
+#undef GetMessage // ignore Windows macro
+
+namespace ra {
+namespace ui {
+namespace win32 {
+
+using ra::ui::viewmodels::MessageBoxIcon;
+using ra::ui::viewmodels::MessageBoxButtons;
+using ra::ui::viewmodels::MessageBoxViewModel;
+
+static HRESULT(WINAPI *pTaskDialog)(HWND hwndParent, HINSTANCE hInstance, PCWSTR pszWindowTitle, PCWSTR pszMainInstruction,
+    PCWSTR pszContent, TASKDIALOG_COMMON_BUTTON_FLAGS dwCommonButtons, PCWSTR pszIcon, int *pnButton) = nullptr;
+
+Dlg_MessageBox::Controller::Controller() noexcept
+{
+    // TaskDialog isn't supported on WinXP, so we have to dynamically find it.
+    auto hDll = LoadLibraryA("comctl32");
+    if (hDll)
+        pTaskDialog = (HRESULT(WINAPI*)(HWND, HINSTANCE, PCWSTR, PCWSTR, PCWSTR, TASKDIALOG_COMMON_BUTTON_FLAGS, PCWSTR, int *))GetProcAddress(hDll, "TaskDialog");
+}
+
+bool Dlg_MessageBox::Controller::IsSupported(const ra::ui::WindowViewModelBase& oViewModel)
+{
+    return (dynamic_cast<const ra::ui::viewmodels::MessageBoxViewModel*>(&oViewModel) != nullptr);
+}
+
+void Dlg_MessageBox::Controller::ShowWindow(ra::ui::WindowViewModelBase& oViewModel)
+{
+    ShowModal(oViewModel);
+}
+
+void Dlg_MessageBox::Controller::ShowModal(ra::ui::WindowViewModelBase& oViewModel)
+{
+    auto& oMessageBoxViewModel = reinterpret_cast<MessageBoxViewModel&>(oViewModel);
+    int nButton = 0;
+
+    if (pTaskDialog == nullptr || oMessageBoxViewModel.GetHeader().empty())
+    {
+        std::wstring sMessage;
+        const std::wstring* pMessage = &oMessageBoxViewModel.GetMessage();
+        if (pTaskDialog == nullptr && !oMessageBoxViewModel.GetHeader().empty())
+        {
+            sMessage.reserve(oMessageBoxViewModel.GetHeader().length() + pMessage->length() + 2);
+            sMessage = oMessageBoxViewModel.GetHeader() + L"\n\n" + *pMessage;
+            pMessage = &sMessage;
+        }
+
+        UINT uType;
+        switch (oMessageBoxViewModel.GetIcon())
+        {
+            default:
+            case MessageBoxIcon::None: uType = 0; break;
+            case MessageBoxIcon::Info: uType = MB_ICONINFORMATION; break;
+            case MessageBoxIcon::Warning: uType = MB_ICONWARNING; break;
+            case MessageBoxIcon::Error: uType = MB_ICONERROR; break;
+        }
+
+        switch (oMessageBoxViewModel.GetButtons())
+        {
+            default:
+            case MessageBoxButtons::OK: uType |= MB_OK; break;
+            case MessageBoxButtons::OKCancel: uType |= MB_OKCANCEL; break;
+            case MessageBoxButtons::YesNo: uType |= MB_YESNO; break;
+            case MessageBoxButtons::YesNoCancel: uType |= MB_YESNOCANCEL; break;
+            case MessageBoxButtons::RetryCancel: uType |= MB_RETRYCANCEL; break;
+        }
+
+        nButton = MessageBoxW(GetActiveWindow(), pMessage->c_str(), oMessageBoxViewModel.GetWindowTitle().c_str(), uType);
```

Comment:
here's the actual Win32 API call. there's another one on line 101 for TaskDialog